validators/cgroup_validator_linux.go

```diff
@@ -66,10 +65,21 @@ func getUnifiedMountpoint(path string) (string, error) {
 		// Example fields: `cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0`.
 		fields := strings.Fields(line)
 		if len(fields) >= 3 {
+			// If default unified mount point is available, return it directly.
+			if fields[1] == defaultUnifiedMountPoint {
+				if fields[2] == "tmpfs" {
+					c, err := os.Open(filepath.Join(defaultUnifiedMountPoint, "cgroup.controllers"))
```

Comment:
this more seems like a line to verify it is v2, not verify it's not v1. I don't personally know of a situation the kernel would return tmpfs but also populate cgroup.controllers

Reply:
is there a common way to check if this is cgroup v1 when the mount type is tmpfs？

Reply:
https://gist.github.com/pacoxu/2768337b62ccc150e8623d25ea890cfc I collected several examples:

- https://gist.github.com/pacoxu/2768337b62ccc150e8623d25ea890cfc?permalink_comment_id=5241477#gistcomment-5241477 is the special one, which is using tmpfs to mount /sys/fs/cgroup and use cgroup v1. But there is a cgroup v2 mountpoint /sys/fs/cgroup/unified.




> this more seems like a line to verify it is v2, not verify it's not v1. I don't personally know of a situation the kernel would return tmpfs but also populate cgroup.controllers


To verify it is not v1, we may check `/sys/fs/cgroup/cpu` or `/sys/fs/cgroup/memory` which is used by kubelet directly.

Reply:
https://github.com/kubernetes/system-validators/pull/43/commits/a520b48ecdd0e63b303e275a9e2d54574518547a is to address this comment. PTAL.

Reply:
Or check `/sys/fs/cgroup/memory` in `/proc/mounts` like below:

```
scanner := bufio.NewScanner(f)
	var cgroupV1MountPoint string
+	var isDefault bool
	for scanner.Scan() {
		line := scanner.Text()
		if !strings.Contains(line, "cgroup") {
			continue
		}
		// Example fields: `cgroup2 /sys/fs/cgroup cgroup2 rw,seclabel,nosuid,nodev,noexec,relatime 0 0`.
		fields := strings.Fields(line)
		if len(fields) >= 3 {
+			if isDefault {
+				if fields[1] == defaultUnifiedMountPoint + "/memory" {
+					return defaultUnifiedMountPoint, fields[2] == "cgroup2", nil
+				}
+				continue
+			}

			// If default unified mount point is available, return it directly.
			if fields[1] == defaultUnifiedMountPoint {
+				isDefault = true
				if fields[2] == "tmpfs" {
+					continue
				}
				return defaultUnifiedMountPoint, fields[2] == "cgroup2", nil
			}
			switch fields[2] {
			case "cgroup2":
				// Return the first cgroups v2 mount point directly.
				return fields[1], true, nil
			case "cgroup":
				// Set the first cgroups v1 mount point only,
				// and continue the loop to find if there is a cgroups v2 mount point.
				if len(cgroupV1MountPoint) == 0 {
					cgroupV1MountPoint = fields[1]
				}
			}
		}
```

The cons here are
- no checking local file